cloudinit/config/cc_grub_dpkg.py

```diff
@@ -43,10 +44,62 @@
 import os
 
 from cloudinit import util
+from cloudinit.util import ProcessExecutionError
 
 distros = ['ubuntu', 'debian']
 
 
+def fetch_idevs(log):
+    """
+    Fetches the /dev/disk/by-id device grub is installed to.
+    Falls back to plain disk name if no by-id entry is present.
+    """
+    disk = ""
+    devices = []
+
+    try:
+        # get the root disk where the /boot directory resides.
+        disk = util.subp(['grub-probe', '-t', 'disk', '/boot'],
+                         capture=True)[0].strip()
+    except ProcessExecutionError as e:
+        # grub-common may not be installed, especially on containers
+        # FileNotFoundError is a nested exception of ProcessExecutionError
+        if isinstance(e.reason, FileNotFoundError):
+            log.debug("'grub-probe' not found in $PATH")
+        # disks from the container host are present in /proc and /sys
+        # which is where grub-probe determines where /boot is.
+        # it then checks for existence in /dev, which fails as host disks
+        # are not exposed to the container.
+        elif "failed to get canonical path" in e.stderr:
+            log.debug("grub-probe 'failed to get canonical path'")
+        else:
+            # something bad has happened, continue to log the error
+            raise
+    except Exception:
+        util.logexc(log, "grub-probe failed to execute for grub-dpkg")
+
+    try:
+        # check if disk exists and use udevadm to fetch symlinks
+        if disk and os.path.exists(disk):
```

Comment:
If grub-probe -t returns OK, but does *not* provide a disk value (which this check seems to think is possible), should we exit earlier and just move this check up after the try/except for grub-probe?

Reply:
If `grub-probe` fails, we'd have `disk = ""` still from line 57. So it at least protects us from that. (I'm not 100% sure I've understood the request, so if this reply doesn't make sense, please do LMK.)

Reply:
I'm suggesting that we do this after the grub-probe call and before we try udev call.

if not disk:
   return []

This handles either scenario I was thinking of:  1) grub-probe fails, we have disks = ""  2) grub-probe returns 0, but produces no standard-output, so disk = ""

If we don't have a disk value, we might as well just return []

Reply:
Aha, right, I see what you mean. Good point, on it.